docs/source/development/controllers.rst

```diff
@@ -0,0 +1,221 @@
+
+For example::
+
+        class MyController(Controller):
+            def control(self):
+                # Get the temperature history for all vials for past hour
+                temp_history = self.evolver.history.get('temp', t_start=time.time() - 3600)
+                for vial in self.vials:
+                    mean_temp = np.mean([i.data[0]['value'] for i in  history.get(name='test').data['test']])
+
+.. note::
+
+    The response from the history server
+    :py:class:`evolver.history.HistoryResult` is amenable to transport over the
+    network, but presently has overhead in working with results in the
+    interpreter (as can be noted in the above example). Functionality like
+    `to_dataframe` or `to_numpy` can be provided in future versions.
+
+
+.. _evolver-hookup:
+
+Evolver hookup and portability
+------------------------------
+
+You might notice that the controller in the example above implicitly requires an
+instance of `evolver.device.Evolver` to be passed into its constructor, in order
+to access the hardware it needs. The framework normally does this for you (it
+will both construct the object and pass in the Evovler), but it may be desired
+to have a controller which depends only on that which is specified in its config
+and takes in hardware explicitly, which can operate independently of the
+application framework.
+
+To do this while also supporting usage within the application, it is recommended
+to specify hardware in the Config which have types accepting either a string or
+hardware class, then a property which dispatches appropriately, for example::
+
+    class MyController(Controller):
+        class Config(Controller.Config):
+            temp_sensor: str|Temperature|ConfigDescriptor
+
+        @property
+        def temp_sensor_hw(self):
+            if isinstance(self.temp_sensor, str):
+                return self.evolver.hardware[self.temp_sensor]
+            return self.temp_sensor
+
+In the above case, when the controller operates within the application, the
+temp_sensor will be specified as the name of the hardware in the evolver
+configuration file, and the controller will use the hardware instance from the
+evolver.
+
+If the controller should be instantiated outside the framework (and without an
+evolver instance), the instantiated Temperature object can be passed in
+directly::
+
+    temp_sensor = Temperature()
+    controller = MyController(temp_sensor=temp_sensor)
+
+Finally, if a ConfigDescriptor (the config object representing how to construct
+the appropriate class) is supplied, it will automatically be processed and
+instantiated::
+
+    descriptor = ConfigDescriptor(classinfo='evolver.hardware.standard.Temperature')
+    controller = MyController(temp_sensor=descriptor)
+
+or with a dict, from create::
+
+    config = {'temp_sensor': {'classinfo': 'evolver.hardware.standard.temperature.Temperature'}}
+    controller = MyController.create(config)
```

Comment:
I think this section needs to be hoisted further up as it shouldn't be an afterthought. Specifying config as a ``dict`` is the principle mechanism for doing so.